aruco_detect/src/map.cpp

```diff
@@ -0,0 +1,415 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer. 
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution. 
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are
+ * those of the authors and should not be interpreted as representing official
+ * policies, either expressed or implied, of the FreeBSD Project.
+ *
+ */
+
+#include <map.h>
+
+#include <string>
+#include <tf2/LinearMath/Vector3.h>
+#include <tf2/LinearMath/Quaternion.h>
+
+#include <std_msgs/String.h>
+#include <std_msgs/ColorRGBA.h>
+#include <geometry_msgs/TransformStamped.h>
+#include <visualization_msgs/Marker.h>
+
+static double deg2rad(double deg)
+{
+    return deg * M_PI / 180.0;
+}
+
+static double rad2deg(double rad)
+{
+    return rad * 180.0 / M_PI;
+}
+
+static double updateVarianceDavid(tf2::Vector3 newMean,
+                                  tf2::Vector3 mean1, double var1,
+                                  tf2::Vector3 mean2, double var2) {
+    //=((2*PI())^0.5)*C3*D3*EXP((((((C2-E2)^2))/(2*C3^2))+(((D2-E2)^2)/(2*(D3^2)))))
+    double d1 = (mean1 - newMean).length2();
+    double d2 = (mean2 - newMean).length2();
+
+    double newVar = sqrt(2.0*M_PI) * var1 * var2 * 
+         exp(((d1 / (2.0*var1)) + d2 / (2.0*var2)));
+
+    if (newVar > 100)
+        newVar = 100;
+    if (newVar < 10e-4)
+        newVar = 10e-4;
+    return newVar;
+}
+static double updateVarianceAlexey(double var1, double var2) {
+    return max(1.0 / (1.0/var1 + 1.0/var2), 1e-6);
+}
+
+static void updateTransform(tf2::Transform &t1, double var1, tf2::Transform &t2, double var2) {
+    tf2::Vector3 o1 = t1.getOrigin();
+    tf2::Vector3 o2 = t2.getOrigin();
+ 
+    t1.setOrigin((var1 * o2 + var2 * o1) / (var1 + var2));
+    
+    tf2::Quaternion q1 = t1.getRotation();
+    tf2::Quaternion q2 = t2.getRotation();
+    t1.setRotation(q1.slerp(q2, var1 / (var1 + var2)).normalize());
```

Comment:
Should the variances be the same? David says that it doesn't really make sense, me might need to do multivariate stuff.